### PR TITLE
Allow deletion of uninstantiated modules whose top level parent import finished. Resolves #2286.

### LINKED ISSUE #2286: Hot-reload fails if exception was thrown during linking of dep modules
### Demonstration

https://codesandbox.io/s/sad-curran-my4dg?file=/index.js

### Expected Behavior

`System.delete()` could delete the parent module.

### Actual Behavior

`System.delete()` could not delete the parent module.

### PR DESCRIPTION
**Pull request by joeldenning**
See https://github.com/systemjs/systemjs/issues/2286#issuecomment-746240996. This test confirms that there is currently a bug. I hope to push a fix in a subsequent commit.

### CODE REVIEW

**Comment by joeldenning** [comment]
I found a fix and pushed it, so this is reviewable now as a potential fix for #2286.

File: src/features/registry.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ systemJSPrototype.delete = function (id) {
   var load = registry[id];
   // in future we can support load.E case by failing load first
   // but that will require TLA callbacks to be implemented
-  if (!load || load.e !== null || load.E)
+  if (!load || load.E)
```
Comment (joeldenning) [inline src/features/registry.js:70]: As explained in https://github.com/systemjs/systemjs/issues/2286#issuecomment-748372786, I think we should allow deletion of modules that have been loaded, but not instantiated. Not sure if this has other negative consequences for other situations.

**Comment by guybedford** [comment]
The `System.delete` logic was written to ensure that it cannot result in "breaking partial state" - that is while a module is instantiating it should not be possible to delete it. Perhaps we can track this by noting the instantiated state eg having that set `load.e = null` or something similar.

We could land this that effectively reverts the previous PR to provide the coherent state checks, but I think seeking to find a comprehensive solution to the partial state problem would be preferable to avoid footguns.

**Comment by joeldenning** [comment]
> The System.delete logic was written to ensure that it cannot result in "breaking partial state" - that is while a module is instantiating it should not be possible to delete it.

As I mentioned in https://github.com/systemjs/systemjs/issues/2286#issuecomment-748372786, there are modules that are inserted into the SystemJS registry which will *never instantiate*. Based on [this code sandbox](https://codesandbox.io/s/hopeful-mestorf-byfu6), it appears that that behavior matches the browser module behavior, so I do not consider the behavior to be a bug. Do you agree that it's not a bug?

If it's not a bug, then I don't think that setting `load.e = null` makes sense, as the module could still later finish instantiation via a top level System.import().

If setting `load.e = null` doesn't make sense, then the question becomes "how should the systemjs delete api treat such modules." Based on your comment, you lean towards not allowing deletion? My concern with that is that users will need to force completion of the instantiation with a top level System.import in order to be able to delete the module, but it will not be obvious that that's what needs to be done.

**Comment by guybedford** [comment]
> Do you agree that it's not a bug?

Yes this is not a bug.

> how should the systemjs delete api treat such modules

The key requirement is that it should not be possible to break an in-progress load that will cause bad state by deleting a module that is currently part of a single top-level instantiation operation.

> My concern with that is that users will need to force completion of the instantiation with a top level System.import in order to be able to delete the module

The module should still be deletable once the top-level instantiation job fails, and that is the fix I think would make the most sense here.

The key distinction here is the top-level instantiation job owning the modules it is acting upon, and noting they are not part of the job.

I'm happy to work on this when I get the time... I just have a lot to do before the end of the year still so likely only late Jan at this point.

**Comment by joeldenning** [comment]
> The module should still be deletable once the top-level instantiation job fails, and that is the fix I think would make the most sense here.

I see, this makes sense, thanks for your guidance here. I'm happy to look into an implementation here, too. Perhaps then System.delete should look at `load.C` instead of `load.e`? Since `C` is the promise for top-level completion?

**Comment by guybedford** [comment]
@joeldenning yes, if only we could synchronously inspect the promise completion we'd be fine :) Perhaps there's a way to extend that check exactly that would work.

**Comment by guybedford** [comment]
Note that `C` is only set on modules _for which top-level jobs are intiated_ (ie entry points). But an approach that sets a `.c = { executing: true }` record that is a single object by reference on each load down the tree would be a nice approach actually.

**Comment by guybedford** [comment]
Actually even easier would just be to reference the top-level load for each submodule - `.P = parentTopLevelLoad`.

The check is then `load.P.e === null`.

**Comment by joeldenning** [comment]
I was coming to a similar conclusion about the need for a new property on the load object, too. I'll give the `load.P.e` check a try.

**Comment by joeldenning** [comment]
I pushed a new commit with the implementation discussed above. I named the property `p` instead of `P`, since the comments in the code indicate that uppercase properties are promises.

File: src/system-core.js [hunk 2]
```diff
@@ -199,7 +199,10 @@ export function getOrCreateLoad (loader, id, firstParentUrl) {
     // On execution, L, I, E cleared
 
     // Promise for top-level completion
-    C: undefined
+    C: undefined,
+
+    // load for firstParent
+    p: loader[REGISTRY][firstParentUrl]
```
Comment (guybedford) [inline src/system-core.js:205]: The concept is correct but unfortunately `firstParentUrl` only refers to the first direct parent, not the top-level entry point.

**Comment by guybedford** [comment]
@joeldenning I pushed a simple approach here which might work. I haven't tested it though, so there may be bugs.

**Comment by guybedford** [comment]
Weirldy the error now seems to be with the name registration of AMD modules! Hopefully we can collaboratively fix, I will likely have to leave it at this for now....

**Comment by joeldenning** [comment]
Thanks for your help with the fix. I found the remaining issue and pushed a fix in https://github.com/systemjs/systemjs/pull/2291/commits/6ecb3dd1907b3b6d187a9b2064e0d0bef1f6af99. I think it was just a typo - the code was calling `postOrderExec()` with the same load argument twice. All tests are now passing. Shall we merge?

File: src/system-core.js [hunk 3]
```diff
@@ -199,18 +197,23 @@ export function getOrCreateLoad (loader, id, firstParentUrl) {
...
 
-function instantiateAll (loader, load, loaded) {
+function instantiateAll (loader, load, parent, loaded) {
   if (!loaded[load.id]) {
     loaded[load.id] = true;
     // load.L may be undefined for already-instantiated
     return Promise.resolve(load.L)
     .then(function () {
+      if (load.e !== null)
+        load.p = parent;
```
Comment (guybedford) [inline src/system-core.js:214]: Just thinking about this a little more carefully, this should probably be:

```suggestion
      if (load.e !== null && (!load.p || load.p.e === null))
        load.p = parent;
```

**Review by guybedford — COMMENTED** [review]
Nice, thanks for spotting the fix here!

I'm going to merge in the suggestion I just made and let's see if the tests still seem to pass (I assume they will).

Then sure let's get this released.

**Comment by joeldenning** [comment]
Thanks I'll merge and release it now.

**Pull request merged**